File: apps/mobile-ios/tromso/tromso/PlatformPreference.swift
//
//  PlatformPreference.swift
//  WallpaperInfoApp
//
//  Created by Brad Park on 4/17/20.
//  Copyright © 2020 Brad Park. All rights reserved.
//

import Foundation

class PlatformPreference {
	
    class func getPreferences() -> WallpaperInfoPreference {
		let defaults = UserDefaults.standard
		
		defaults.register(defaults: ["root_path" : "BP Photo" ])
		defaults.register(defaults: ["theme" : String(Theme.default1.rawValue) ])
		defaults.register(defaults: ["custom_root" : ThemeInfo._default_custom_root ])
		defaults.register(defaults: ["custom_allow" : ThemeInfo._default_custom_allow ])
		defaults.register(defaults: ["custom_filter" : ThemeInfo._default_custom_filter ])
		defaults.register(defaults: ["pause" : String(false) ])
        defaults.register(defaults: ["interval" : String(WallpaperServiceInfo._defaultInterval) ])
        defaults.register(defaults: ["offline_mode" : String(true) ])
        defaults.register(defaults: ["last_sync_time" : "" ])

        return WallpaperInfoPreference(root_path: defaults.string(forKey: "root_path")!,
									   theme: Theme(rawValue: Int(defaults.string(forKey: "theme")!)!)!,
									   custom_root: defaults.string(forKey: "custom_root")!,
									   custom_allow: defaults.string(forKey: "custom_allow")!,
									   custom_filter: defaults.string(forKey: "custom_filter")!,
									   pause: Bool(defaults.string(forKey: "pause")!)!,
									   interval: Int(defaults.string(forKey: "interval")!)!,
                                       offline_mode: Bool(defaults.string(forKey: "offline_mode")!)!,
                                       last_sync_time: defaults.string(forKey: "last_sync_time")!)
	}
	
	class func setPreference(key:String, value:String) {
		let defaults = UserDefaults.standard
		defaults.set(value, forKey: key)
	}
}

public struct WallpaperInfoPreference {
    var root_path: String
	var theme: Theme
	var custom_root:String
	var custom_allow:String
    var custom_filter: String
    var pause: Bool
    var interval: Int
    var offline_mode: Bool
    var last_sync_time: String
}

File: apps/mobile-android/wallpaperinfo/app/src/main/java/com/thirdwavesoft/wallpaperinfo/PlatformPreference.java
package com.thirdwavesoft.wallpaperinfo;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

public class PlatformPreference {

    synchronized static WallpaperInfoPreference getPreferences(Context context) {
        WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
        WallpaperInfoPreference wpref = new WallpaperInfoPreference();
        SharedPreferences pref = context.getSharedPreferences(wallpaperInfoPref, Context.MODE_PRIVATE); // 8
        wpref.root_path = pref.getString("root_path", null);
        wpref.theme = Theme.rawValue(Integer.valueOf(pref.getString("theme", (new Integer(Theme.default1.intValue()).toString()))));
        wpref.custom_root = pref.getString("custom_root", ThemeInfo._default_custom_root);
        wpref.custom_allow = pref.getString("custom_allow", ThemeInfo._default_custom_allow);
        wpref.custom_filter = pref.getString("custom_filter", ThemeInfo._default_custom_filter);
        wpref.pause = new Boolean(pref.getString("pause", (new Boolean(false)).toString()));
        wpref.interval = Integer.valueOf(pref.getString("interval", (new Integer(wpsi._defaultInterval).toString())));
        wpref.saverTime = Integer.valueOf(pref.getString("saver_time", (new Integer(wpsi._defaultSaverTime).toString())));
        return wpref;
    }

    synchronized static void setPreference(Context context, String key, String value) {
        if (context != null) {
            SharedPreferences pref = context.getSharedPreferences(wallpaperInfoPref, Context.MODE_PRIVATE); // 0
            SharedPreferences.Editor editor = pref.edit();
            editor.putString(key, value);
            editor.commit();
        } else {
            Log.e(_TAG, "setWallpaperInfoProperty failed due to no conext.");
        }
    }

    // default values
    private final static String _TAG = "PlatformPreference";
    private final static String wallpaperInfoPref = "prefWallpaperInfoPref";

}

class WallpaperInfoPreference {
    public String root_path;
    public Theme theme;
    public String custom_root;
    public String  custom_allow;
    public String custom_filter;
    public boolean pause;
    public int interval;
    public int saverTime;
}